src/librustc_mir/interpret/intern.rs

```diff
@@ -319,10 +327,23 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
-            match base_intern_mode {
-                InternMode::Static => {}
-                InternMode::Const | InternMode::ConstBase => {
-                    // If it's not a static, it *must* be immutable.
+            match intern_kind {
+                // Mutable statics may contain mutable allocations even behind relocations
+                InternKind::Static(hir::Mutability::Mut) => {}
+                // Once we get heap allocations we need to revisit whether immutable statics can
+                // refer to mutable (e.g. via interior mutability) allocations.
+                InternKind::Static(hir::Mutability::Not) => {
+                    alloc.mutability = Mutability::Not;
```

Comment:
Why is this correct, given things like `static FOO: AtomicUsize`, which is mutable? The old code didn't do anything like this, did it?

Reply:
hmm... I didn't think about that, reverting to the previous logic

Reply:
I adjusted the comment, That situation can never occur, the alloc ids we look at here are never the base allocation's alloc id

Reply:
I don't see why these all have to be immutable. For constants that makes sense, but why for statics?

Reply:
the only way you can get a relocation in statics is by using promotion, and promotion only works for immutable references to frozen types. So no matter what you do with that pointer you get, you will never be able to write to the memory without UB

Reply:
So does promotion still do "no `StorageDead`"?

Reply:
well... `StorageDead` instructions just aren't moved to the promoted MIR, so... 

> does promotion still do "no StorageDead"?

yes, they have always done this

Reply:
Well you told me recently promotion no longer does this, and works by actually extracting separate MIR the way it does from run-time code. Now you are saying the opposite. I am very confused.

Reply:
promotion in constants and statics used to just drop the `StorageDead` markers. Now it actually extracts a real promoted, but that promoted may contain interior promoteds that again use the `StorageDead` trick.

So in `const X: &'static i32 = &5;` you get

```rust
const PROMOTED: &'static i32 = {
    let tmp = 5;
    &tmp // no storage dead for tmp
};
const X: &'static i32 = PROMOTED;
```

and in `const Y: &'static &'static i32 = &&5;` you get

```rust
const PROMOTED: &'static i32 = {
    let a = 5;
    let b = &a;
    &b // no storage dead for a and b
};
const X: &'static &'static i32 = PROMOTED;
```

Reply:
> Now it actually extracts a real promoted, but that promoted may contain interior promoteds that again use the StorageDead trick.

Right, so promoteds *inside promoteds* still just kill `StorageDead`. But the branch this discussion about is `InternKind::Static`, not a promoted, so that part is rrelevant.

This is a static, and thus shouldn't be able to have any inner relocations, those would all have been promoted already. What am I missing?

Reply:
The new arm for `Static` here isn't critical for the bugfix, right? Only the `Promoted` thing is. So in the interest of getting the regression fixed ASAP, I suggest removing this part from this PR and having that discussion in a new PR.

Reply:
Ah, this entire match is only for relocations outside of references. So your example `static FOO: AtomicUsize` wouldn't even hit this branch, you'd need `static FOO: *const AtomicUsize = &AtomicUsize::new(42);`, which... is not problematic at all, so the code is indeed wrong.

> The new arm for Static here isn't critical for the bugfix, right? Only the Promoted thing is. So in the interest of getting the regression fixed ASAP, I suggest removing this part from this PR and having that discussion in a new PR.

Yes, I'll remove it